File: src/main/java/me/testaccount666/serversystem/ServiceRegistry.java
package me.testaccount666.serversystem;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public final class ServiceRegistry {
    private final Map<Class<?>, Object> _services = new ConcurrentHashMap<>();

    <T> T registerService(Class<T> type, T service) {
        if (hasService(type)) throw new IllegalStateException("Service already registered: ${type.getName()}");

        _services.put(type, service);

        return service;
    }

    public <T> T getService(Class<T> type) {
        return type.cast(_services.get(type));
    }

    public <T> Optional<T> getServiceOptional(Class<T> type) {
        return Optional.ofNullable(_services.get(type)).map(type::cast);
    }

    public boolean hasService(Class<?> type) {
        return getServiceOptional(type).isPresent();
    }

    void clearServices() {
        _services.clear();
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/ServiceRegistry.kt
package me.testaccount666.serversystem

import java.util.concurrent.ConcurrentHashMap

class ServiceRegistry {
    private val _services: MutableMap<Class<*>, Any> = ConcurrentHashMap()

    @Synchronized
    fun <T : Any> registerService(type: Class<T>, service: T): T {
        check(!hasService(type)) { "Service already registered: ${type.name}" }

        _services[type] = service
        return service
    }

    fun <T : Any> getService(type: Class<T>): T = type.cast(_services[type])

    fun <T : Any> getServiceOrNull(type: Class<T>): T? = _services[type]?.let { type.cast(it) }

    fun <T : Any> hasService(type: Class<T>): Boolean = type in _services

    @Synchronized
    fun clearServices() = _services.clear()

    inline fun <reified T : Any> registerService(service: T): T = registerService(T::class.java, service)
    inline fun <reified T : Any> getService(): T = getService(T::class.java)
    inline fun <reified T : Any> getServiceOrNull(): T? = getServiceOrNull(T::class.java)
    inline fun <reified T : Any> hasService(): Boolean = hasService(T::class.java)

}